### PR TITLE
Expanding multiple leaves at a time for MCTS

### PR DESCRIPTION
**Pull request by emailweixu**
MCTS is slow because it needs to do many rounds of tree expansion. This PR tries to alleviate this my doing multiple expansions simultaneously.

Also added more summaries for MCTS

Test plan:
1. mcts_algorithm_test.py
2. with muzero_pendulum_conf, got similar performance with num_parallel_sim=1 or 2
  num_parallel_sims=1: mean=157.0, std=24.0
  num_parallel_sims=2: mean=164.2, std=36.8

### CODE REVIEW

File: alf/algorithms/mcts_algorithm.py [hunk 1]
```diff
@@ -279,6 +313,11 @@ def __init__(
                 using reverse KL divergence will be used for tree search.
             learn_with_exploration_policy (bool): If True, a policy calculated
                 using reverse KL divergence will be used for learning.
+            max_unroll_length (int): maximal allowed unroll steps when building
+                the search tree.
+            num_parallel_sims (int): expanding so many leaves at a time for one
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py:326]: Suggest adding a note on that `num_simulations % num_parallel_sims = 0`. This also implies a bit on how `num_parallel_sims` is being used and can help the reader, and guide the user to set proper value.

File: alf/algorithms/mcts_algorithm.py [hunk 2]
```diff
@@ -252,6 +283,9 @@ def __init__(
                 where ``steps`` is a vector representing the number of steps in
                 the games. And it is expected to return a float vector of the same
                 shape as ``steps``.
+            learn_policy_temperature (float): transform the policy p found by
+                MCTS by :math:`p^{1/learn_policy_temperature} / Z` as policy
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py:292]: It does not immediately occur to me that `Z` is a normalization factor, probably worth noting it.

File: alf/algorithms/mcts_algorithm.py [hunk 3]
```diff
@@ -194,6 +200,27 @@ class MCTSAlgorithm(OffPolicyAlgorithm):
...
+      When contructing the search paths, the i-th search path is based on the
+      i-th best child of each node.
+
+    * search_with_exploration_policy=False. The best child is same the case k=1.
+      The second best child is found by assuming the visit count of the best
+      child and the parent are increased by 1 and applying the UCB criterion
+      again. This is repeated k times to get k best children. During the process
+      of traversing from the root to contruct k search paths, if the current end
+      node of several (let's say k') paths are the same, we will use best k'
+      children of that node to extend the paths.
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py]: When I reading this paragraph without reading the code yet, I had a few questions:

1. Is this procedure seems "not parallel" across `k`?
2. "second best" is a bit ambiguous here, it can mean "the one that is only worse to the best one" or "the best that is picked in the second sampling".
3. Had some hard time understanding this "if the current end node of several (let's say k') paths are the same, we will use best k' children of that node to extend the paths". Some Ascii graph might help. Also I am confused whether this behavior is
   * a special treatment or,
   * is implied from the procedure but highlighted

File: alf/algorithms/mcts_algorithm.py [hunk 4]
```diff
@@ -279,6 +313,11 @@ def __init__(
                 using reverse KL divergence will be used for tree search.
             learn_with_exploration_policy (bool): If True, a policy calculated
                 using reverse KL divergence will be used for learning.
+            max_unroll_length (int): maximal allowed unroll steps when building
+                the search tree.
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py]: Would be better if we state the connection between `max_unroll_length` and the maximum tree depth. This would help clarify to the reader how it affects the tree search.

File: alf/algorithms/mcts_algorithm.py [hunk 5]
```diff
@@ -453,9 +586,13 @@ def _build_tree2(self, trees, to_plays):
         trees.value_sum[roots] += branch_factor * v
         self._update_best_child(trees, roots)
 
+        max_depth = torch.zeros(trees.B.shape[0], dtype=torch.int64)
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py:597]: There is already `self._max_depth` and this naming seems to have a conflict. Would something like `longest_path` (and `avg_path_length`) or `deepest` work?

File: alf/algorithms/mcts_algorithm.py [hunk 6]
```diff
@@ -505,19 +646,59 @@ def _search(self, trees, to_plays):
         children_index = trees.children_index
         best_child_index = trees.best_child_index
         game_over = trees.game_over
+        psims = self._num_parallel_sims
+        parallel = self._parallel
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py]: This shorthand seems unnecessary.

File: alf/algorithms/mcts_algorithm.py [hunk 7]
```diff
@@ -505,19 +646,59 @@ def _search(self, trees, to_plays):
...
         search_paths = []
         B = trees.B
         nodes = trees.root_indices
-        path_lengths = torch.ones_like(B)
+        calc_i = False
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py]: I think it is worth explaining what `i` is here, since this part of code is already hundreds lines away from where "parallel sims" is explained. Or give a pointer to where "parallel sims" is explained above (or below).

File: alf/algorithms/mcts_algorithm.py [hunk 8]
```diff
@@ -505,19 +646,59 @@ def _search(self, trees, to_plays):
...
+        if parallel:
+            B = B.unsqueeze(-1)
+            nodes = trees.root_indices.unsqueeze(-1).expand(-1, psims)
+            if self._search_with_exploration_policy:
+                i = (torch.arange(psims), )
+            else:
+                calc_i = True
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py:672]: In this case `i` is not initialized (I know this is not c++, but habits :) ) - Maybe we can just put the `i = ()` in the else branch right before the `if parallel` and remove the `else` branch below?

File: alf/algorithms/mcts_algorithm.py [hunk 9]
```diff
@@ -505,19 +646,59 @@ def _search(self, trees, to_plays):
...
+            if calc_i:
+                # If several nodes are same, we should select actions from different
+                # entries in best_child_index. So we need to find out which entry
+                # of best_child_index a node should use. For search_with_exploration_policy,
+                # the entries in best_child_index are sampled independently from
+                # the calculated policy, so it is ok to use i'th best_child_index
+                # for i-th node. But for UCB based selection, we want to use
+                # the top entries in best_child_index when ever they are not used.
+                #
+                # For x>y, i[b, x, y] indicates where nodes[b, x] and nodes[b, y] are same
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py]: It is a bit confusing here when `i` is explained as 3 dimensional here but becomes (a tuple of) 2-d tensors after the `sum` below. Shall we use a different variable?

File: alf/algorithms/mcts_algorithm.py [hunk 10]
```diff
@@ -505,19 +646,59 @@ def _search(self, trees, to_plays):
...
+                # For example, suppose nodes[0] = [2,3,2,3,3], i[0] will be
+                # [[0, 0, 0, 0, 0],
+                #  [0, 0, 0, 0, 0],
+                #  [1, 0, 0, 0, 0],
+                #  [0, 1, 0, 0, 0],
+                #  [0, 1, 0, 1, 0]]
+                # [B, psims, psims]. `None` stands for unsqueezing at that dim.
+                i = (nodes[:, :, None] == nodes[:, None, :]).tril(diagonal=-1)
+                # With the above example, i[0] is [0, 0, 1, 1, 2]
+                # [B, psims]
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py:719]: A further example on how this is affecting the `torch.where` below would be very helpful. I have to admit that this alone takes quite some time to digest.

**Review by breakds — COMMENTED** [review]
First half of review comments.

File: alf/algorithms/mcts_algorithm.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:326]: added.

File: alf/algorithms/mcts_algorithm.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: added.

File: alf/algorithms/mcts_algorithm.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:292]: added.

File: alf/algorithms/mcts_algorithm.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: Added explanation for 2 and 3. 
For 1,  although the part for calculating the k children is not parallel, its time is little compared to other part (which are parallel).

File: alf/algorithms/mcts_algorithm.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:597]: changed self._max_depth to self._max_allowed_depth

File: alf/algorithms/mcts_algorithm.py (hunk 6 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: fixed

File: alf/algorithms/mcts_algorithm.py (hunk 7 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: added.

File: alf/algorithms/mcts_algorithm.py (hunk 8 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:672]: Doesn't seem to be a problem to me. After all, this is python. And that extra step of i=() do incur a small cost.

File: alf/algorithms/mcts_algorithm.py (hunk 10 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:719]: Added comment about i at line 665.

File: alf/algorithms/mcts_algorithm.py (hunk 9 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: changed to `same_node`

File: alf/algorithms/mcts_algorithm.py [hunk 11]
```diff
@@ -508,19 +654,63 @@ def _search(self, trees, to_plays):
...
+                # For example, suppose nodes[0] = [2,3,2,3,3], same[0] will be
+                # [[0, 0, 0, 0, 0],
+                #  [0, 0, 0, 0, 0],
+                #  [1, 0, 0, 0, 0],
+                #  [0, 1, 0, 0, 0],
+                #  [0, 1, 0, 1, 0]]
+                # [B, psims, psims]. `None` stands for unsqueezing at that dim.
+                same_node = (nodes[:, :, None] == nodes[:, None, :]).tril(
+                    diagonal=-1)
+                # With the above example, i[0] is [0, 0, 1, 1, 2]
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py]: I think it would be better to also explain the meaning of the result in `i` w.r.t. the example, since it took me a while to fully under stand what this is doing.

> suppose nodes[0] = [2, 3, 2, 3, 3], it means that the current 5 parallel searches on tree 0 are at tree 0's node 2, node 3, node 2, node 3 and node 3 respectively. According to our algorithm, even under USB-based selection strategy, we want the 3 searches at node 3 to go to different "next nodes" (i.e. the top 3 best children) respectively. The same goes for the 2 searches at node 2 (they should go to the top 2 best children respectively). Therefore we want to use `i` to map nodes index [2, 3, 2, 3, 3] to children index [0, 0, 1, 1, 2].
> ...

File: alf/algorithms/mcts_algorithm.py [hunk 12]
```diff
@@ -420,34 +483,109 @@ def _build_tree1(self, trees, to_plays):
...
+        B = trees.B
+        if parallel:
+            B = B.unsqueeze(-1)
+            i = (torch.arange(psims), )
+        else:
+            i = ()
+        max_depth = torch.zeros(B.shape[0], dtype=torch.int64)
+        avg_depth = 0.
+        new_nodes = []
+        for sim in range(sim0, self._num_simulations + sim0, psims):
```
Comment (breakds) [inline alf/algorithms/mcts_algorithm.py:497]: Suggest renaming the loop variable `sim` to `starting_sim_index`, as the name `sim` will be used below for the variable that holdes `[start_sim_index, start_sim_index + 1, ..., start_sim_index + psims - 1]`

Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:497]: For psims=1, `sim`  is still sim.

File: alf/algorithms/mcts_algorithm.py (hunk 11 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: updated

File: alf/algorithms/mcts_algorithm.py (hunk 8 above)
Comment (breakds) [reply alf/algorithms/mcts_algorithm.py:672]: Sounds good.

**Review by breakds — APPROVED** [review]
